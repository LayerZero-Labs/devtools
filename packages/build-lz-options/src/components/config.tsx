import React from "react";
import { Box, Text } from "ink";
import {
  OptionType1Summary,
  OptionTypeInput,
  OptionType2Summary,
} from "@/types";

interface Props {
  value: OptionTypeInput;
}

interface OptionType1Props {
  props: OptionType1Summary;
}

interface OptionType2Props {
  props: OptionType2Summary;
}

export const ConfigSummary: React.FC<Props> = ({ value }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Creating LayerZero options <Text bold>{value.type.label}</Text>
      </Text>
    </Box>
  );
};

export const Option1Summary: React.FC<OptionType1Props> = ({ props }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Gas Limit: <Text bold>{props.gasLimit}</Text>
      </Text>
    </Box>
  );
};

export const Option2Summary: React.FC<OptionType2Props> = ({ props }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Gas Limit: <Text bold>{props.gasLimit}</Text>
      </Text>
      <Text>
        Native Drop Amount: <Text bold>{props.nativeDropAmount}</Text>
      </Text>
      <Text>
        Native Drop Address: <Text bold>{props.nativeDropAddress}</Text>
      </Text>
    </Box>
  );
};
