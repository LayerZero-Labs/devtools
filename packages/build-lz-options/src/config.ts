import type { OptionType } from '@/types'
import { ExecutorOptionType, WorkerId } from '@layerzerolabs/lz-v2-utilities'

/**
 * Supported Option Types.
 */
export const OPTION_TYPES: OptionType[] = [
    {
        // TODO: use OptionType.TYPE_1 once exported from lz-v2-utility
        id: '1',
        label: '1: gas for remote execution',
    },
    {
        // TODO: use OptionType.TYPE_2 once exported from lz-v2-utility
        id: '2',
        label: '2: gas for remote execution and native drop',
    },
    {
        // TODO: use OptionType.TYPE_3 once exported from lz-v2-utility
        id: '3',
        label: '3: options builder (EndpointV2 only)',
    },
]

/**
 * Supported Executor Option Types.
 */
export const EXECUTOR_OPTION_TYPE = [
    {
        id: ExecutorOptionType.LZ_RECEIVE,
        label: '1: lzReceive',
    },
    {
        id: ExecutorOptionType.NATIVE_DROP,
        label: '2: nativeDrop',
    },
    {
        id: ExecutorOptionType.COMPOSE,
        label: '3: compose',
    },
    {
        id: ExecutorOptionType.ORDERED,
        label: '4: ordered',
    },
]

/**
 * Supported Worker Types.
 */
export const WORKER_TYPE = [
    {
        id: WorkerId.EXECUTOR,
        label: '1 Executor',
    },
    {
        id: WorkerId.VERIFIER,
        label: '2 Verifier',
    },
]
