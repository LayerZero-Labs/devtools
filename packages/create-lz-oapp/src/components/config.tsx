import React from "react";
import { Box, Text } from "ink";
import { Config } from "@/types.js";
import { resolve } from "path";

interface Props {
  value: Config;
}

export const ConfigSummary: React.FC<Props> = ({ value }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Will create a project in{" "}
        <Text bold>{value.destination || "current"}</Text> directory (
        <Text bold>{resolve(value.destination)}</Text>)
      </Text>
      <Text>
        Will use the <Text bold>{value.example.label}</Text> example
      </Text>
      <Text>
        Will use <Text bold>{value.packageManager.label}</Text> to install
        dependencies
      </Text>
    </Box>
  );
};
