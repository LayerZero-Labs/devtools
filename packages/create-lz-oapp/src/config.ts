import type { Example, PackageManager } from '@/types.js'

/**
 * To enable example development in a custom repository
 * we open the repository URL field to be taken from the environment
 */
const repository = process.env['LAYERZERO_EXAMPLES_REPOSITORY_URL'] || 'git@github.com:LayerZero-Labs/lz-utils'

/**
 * To enable example development in a custom branch,
 * we open up the ref field to be taken from the environment
 *
 * `LAYERZERO_EXAMPLES_REPOSITORY_REF` can then be set to something like `#develop` or `#my-custom-branch`
 * to take the examples from a tag, a branch or a commit hash
 */
const ref = process.env['LAYERZERO_EXAMPLES_REPOSITORY_REF'] || ''

export const EXAMPLES: Example[] = [
    {
        id: 'oft',
        label: 'OFT',
        repository,
        directory: 'examples/oft',
        ref,
    },
    {
        id: 'oapp',
        label: 'OApp',
        repository,
        directory: 'examples/oapp',
        ref,
    },
]

export const PACKAGE_MANAGERS: PackageManager[] = [
    {
        command: 'npm',
        label: 'npm',
    },
    {
        command: 'yarn',
        label: 'yarn',
    },
    {
        command: 'pnpm',
        label: 'pnpm',
    },
    {
        command: 'bun',
        label: 'bun',
    },
]
