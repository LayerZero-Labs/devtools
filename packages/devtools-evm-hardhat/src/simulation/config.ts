import type { HardhatConfig, HttpNetworkConfig, NetworkConfig } from 'hardhat/types'
import type { SimulationConfig, SimulationUserConfig } from './types'
import { resolve } from 'path'
import { AnvilOptions } from '@layerzerolabs/devtools-evm'
import { pipe } from 'fp-ts/lib/function'
import * as R from 'fp-ts/Record'

/**
 * Turns `SimulationUserConfig` into `SimulationConfig` by supplying defaults
 *
 * @param {SimulationUserConfig} userConfig
 * @param {HardhatConfig} hardhatConfig
 * @returns {SimulationConfig}
 */
export const resolveSimulationConfig = (
    userConfig: SimulationUserConfig,
    hardhatConfig: HardhatConfig
): SimulationConfig => ({
    port: userConfig.port ?? 8545,
    directory: resolve(hardhatConfig.paths.root, userConfig.directory ?? '.layerzero'),
    overwriteAccounts: userConfig.overwriteAccounts ?? true,
    anvil: {
        // For now we'll hardcode the mnemonic we'll use to seed the accounts on the simulation networks
        mnemonic: 'test test test test test test test test test test test junk',
        ...userConfig.anvil,
        // The host and port need to always point to 0.0.0.0:8545
        // since anvil runs in the container that exposes this port on 0.0.0.0
        host: '0.0.0.0',
        port: 8545,
    },
})

/**
 * Takes a portion of the hardhat networks config and turns the values into `AnvilOptions`
 * to be used by a simulation network container.
 *
 * This is used in the simulation generation phase where the compose files
 * for networks defined in hardhat config are being generated.
 *
 * The `Record<string, NetworkConfig>` is used instead of `NetworksConfig` type from hardhat
 * to avoid any type issues with special keys like `localhost` or `hardhat`.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, NetworkConfig>} networksConfig
 * @returns {Record<string, AnvilOptions>} Object with the same keys as the networks config, with values mapped to `AnvilOptions`
 */
export const getAnvilOptionsFromHardhatNetworks = (
    config: SimulationConfig,
    networksConfig: Record<string, NetworkConfig>
): Record<string, AnvilOptions> =>
    pipe(
        networksConfig,
        // We want to drop all the networks that don't have URLs
        R.filter(isHttpNetworkConfig),
        // And map the network configs into AnvilOptions
        R.map(
            (networkConfig: HttpNetworkConfig): AnvilOptions => ({
                ...config.anvil,
                forkUrl: networkConfig.url,
            })
        )
    )

/**
 * Returns with overrides for hardhat network configuration.
 *
 * This, when merged with a hardhat config, will redirect the network calls to the local EVM nodes.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, NetworkConfig>} networksConfig
 * @returns {Record<string, NetworkConfig>}
 */
export const getHardhatNetworkOverrides = (
    config: SimulationConfig,
    networksConfig: Record<string, NetworkConfig>
): Record<string, NetworkConfig> =>
    pipe(
        networksConfig,
        // We want to drop all the networks that don't have URLs
        R.filter(isHttpNetworkConfig),
        // We'll take the existing network configs and point them to our RPC proxy
        //
        // It's important that these configs are not saved to filesystem as they might contain
        // sensitive data (and forgetting to ignore these files in git could lead to security breaches)
        R.mapWithIndex(
            (networkName, networkConfig): NetworkConfig => ({
                ...networkConfig,
                // We want to redirect this network to the local proxy
                //
                // This is the nginx server listening on the port we configured in the simulation configuration
                url: new URL(networkName, `http://localhost:${config.port}`).toString(),
                accounts: config.overwriteAccounts
                    ? // When overwriting accounts, all the accounts will be switched to the anvil config
                      // (or reasonable defaults if not provided)
                      {
                          mnemonic: config.anvil.mnemonic,
                          // These need to be defaulted to the anvil options
                          // (or the anvil defaults)
                          //
                          // See https://book.getfoundry.sh/reference/cli/anvil for anvil defaults
                          count: config.anvil.count ?? 10,
                          path: config.anvil.derivationPath ?? "m/44'/60'/0'/0/",
                          // These will be hardcoded for now as anvil does not support setting these
                          initialIndex: 0,
                          passphrase: '',
                      }
                    : networkConfig.accounts,
            })
        )
    )

/**
 * Helper utility to pick network configs by their names / object keys.
 *
 * Similar to TypeScript `Pick` helper type, but in runtime.
 *
 * @param {string[]} networks List of networks to pick
 */
export const pickNetworkConfigs = (networks: string[]) =>
    R.filterWithIndex<string, NetworkConfig>((networkName: string) => networks.includes(networkName))

/**
 * Little helper utility that checks whether a network config is not a hardhat network config
 *
 * @param {NetworkConfig} networkConfig
 * @returns {boolean}
 */
const isHttpNetworkConfig = (networkConfig: NetworkConfig): networkConfig is HttpNetworkConfig =>
    'url' in networkConfig && typeof networkConfig.url === 'string'
