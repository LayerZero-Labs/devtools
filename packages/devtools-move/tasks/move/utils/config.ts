import fs from 'fs'
import path from 'path'

import { Account, Ed25519PrivateKey } from '@aptos-labs/ts-sdk'
import YAML from 'yaml'
import { OAppOmniGraphHardhat, OmniPointHardhat } from '@layerzerolabs/toolbox-hardhat'
import { getNetworkForChainId } from '@layerzerolabs/lz-definitions'

type AptosYamlConfig = {
    profiles: {
        default: {
            network: string
            private_key: string
            public_key: string
            account: string
            rest_url: string
            faucet_url?: string
        }
    }
}

export async function getLzConfig(configPath: string): Promise<OAppOmniGraphHardhat> {
    const lzConfigPath = path.resolve(path.join(process.cwd(), configPath))
    const lzConfigFile = await import(lzConfigPath)
    const lzConfig = lzConfigFile.default
    return lzConfig
}

export function getMoveVMContracts(lzConfig: OAppOmniGraphHardhat): OmniPointHardhat[] {
    const contracts = []
    for (const entry of lzConfig.contracts) {
        const chainName = getNetworkForChainId(entry.contract.eid).chainName
        if (chainName === 'aptos' || chainName === 'initia' || chainName === 'movement') {
            contracts.push(entry.contract)
        }
    }
    return contracts
}

export async function loadAptosYamlConfig(_rootDir: string = process.cwd()): Promise<AptosYamlConfig> {
    const configPath = path.resolve(path.join(_rootDir, '.aptos/config.yaml'))

    if (!fs.existsSync(configPath)) {
        throw new Error(
            `Aptos config file not found at ${configPath}.\n\n\tPlease run "aptos init" to initialize your project.\n`
        )
    }

    const file = fs.readFileSync(configPath, 'utf8')
    const config = YAML.parse(file) as AptosYamlConfig
    return config
}

export function createAccountFromPrivateKey(privateKey: string, account_address: string): Account {
    return Account.fromPrivateKey({
        privateKey: new Ed25519PrivateKey(privateKey),
        address: account_address,
    })
}

export function getNamedAddresses(networkType: string): string {
    const addressesPath = path.join(__dirname, './deploymentAddresses.json')
    const addresses = JSON.parse(fs.readFileSync(addressesPath, 'utf8'))
    const networkAddresses = addresses[`${networkType}-addresses`]

    return Object.entries(networkAddresses)
        .map(([name, addr]) => `${name}=${addr}`)
        .join(',')
}
