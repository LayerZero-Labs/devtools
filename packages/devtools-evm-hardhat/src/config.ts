import 'hardhat-deploy/dist/src/type-extensions'

import { endpointIdToNetwork } from '@layerzerolabs/lz-definitions'
import { HardhatUserConfig } from 'hardhat/types'
import { join, dirname } from 'path'
import { createModuleLogger } from '@layerzerolabs/io-devtools'

const resolvePackageDirectory = (packageName: string): string => {
    // The tricky bit here is the fact that if we resolve packages by their package name,
    // we might be pointed to a file in some dist directory - node will just pick up the `main`
    // entry in package.json and point us there
    //
    // So in order to get a stable path we choose package.json, pretty solid choice
    const packageJsonName = join(packageName, 'package.json')
    // We now resolve the path to package.json
    const packageJsonPath = require.resolve(packageJsonName)
    // And return its directory
    return dirname(packageJsonPath)
}

/**
 * Helper utility that adds external deployment paths for all LayzerZero enabled networks.
 * This will make LayerZero contracts available in your deploy scripts and tasks.
 *
 * ```
 * // hardhat.config.ts
 * import { EndpointId } from "@layerzerolabs/lz-definitions"
 *
 * const config: HardhatUserConfig = {
 *   networks: {
 *     arbitrum: {
 *       eid: EndpointId.ARBITRUM_MAINNET
 *     },
 *     fuji: {
 *       eid: EndpointId.AVALANCHE_TESTNET
 *     }
 *   }
 * }
 *
 * export default withLayerZeroDeployments("@layerzerolabs/lz-evm-sdk-v1")
 * ```
 *
 * @param packageNames `string[]` List of @layerzerolabs package names that contain deployments directory
 *
 * @returns `<THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig` Hardhat config decorator
 */
export const withLayerZeroDeployments = (...packageNames: string[]) => {
    const resolvedDeploymentsDirectories = packageNames
        // The first thing we do is we resolve the paths to LayerZero packages
        .map(resolvePackageDirectory)
        // Then navigate to the deployments folder
        .map((resolvedPackagePath) => join(resolvedPackagePath, 'deployments'))

    // We return a function that will enrich hardhat config with the external deployments configuration
    //
    // This is a pretty standard way of enriching configuration files that leads to quite nice consumer code
    return <THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig => ({
        ...config,
        external: {
            ...config.external,
            // Now for the meat of the operation, we'll enrich the external.deployments object
            deployments: Object.fromEntries(
                Object.entries(config.networks ?? {}).flatMap(([networkName, networkConfig]) => {
                    const eid = networkConfig?.eid
                    const networkLogger = createModuleLogger(networkName)

                    // Let's first check whether eid is defined on the network config
                    if (eid == null) {
                        networkLogger.debug(
                            'Endpoint ID not specified in hardhat config, skipping external deployment configuration'
                        )

                        return []
                    }

                    try {
                        // This operation is unsafe and can throw - let's make sure we don't explode with some unreadable error
                        const layerZeroNetworkName = endpointIdToNetwork(eid)
                        const layerZeroNetworkDeploymentsDirectories = resolvedDeploymentsDirectories.map(
                            (deploymentsDirectory) => join(deploymentsDirectory, layerZeroNetworkName)
                        )

                        return [
                            [
                                // The external deployments object is keyed by local network names
                                // which do not necessarily match the LayerZero ones
                                networkName,
                                // And its values are arrays of filesystem paths referring to individual network deployment directories
                                Array.from(
                                    // Since we want the paths to be unique, we'll put everything we have into a Set, then convert back to array
                                    new Set([
                                        // These are the external deployments already configured
                                        ...(config.external?.deployments?.[networkName] ?? []),
                                        // And these are the new ones
                                        ...layerZeroNetworkDeploymentsDirectories,
                                    ])
                                ),
                            ],
                        ]
                    } catch (error) {
                        networkLogger.error(
                            `Invalid endpoint ID specified in hardhat config (${eid}), skipping external deployment configuration`
                        )

                        return []
                    }
                })
            ),
        },
    })
}

/**
 * Helper utility that appends external artifacts directories
 * to existing hadrhat config
 *
 * ```
 * // hardhat.config.ts
 * import { EndpointId } from "@layerzerolabs/lz-definitions"
 *
 * const config: HardhatUserConfig = {
 *   networks: {
 *     arbitrum: {
 *       eid: EndpointId.ARBITRUM_MAINNET
 *     },
 *     fuji: {
 *       eid: EndpointId.AVALANCHE_TESTNET
 *     }
 *   }
 * }
 *
 * export default withLayerZeroArtifacts("@layerzerolabs/lz-evm-sdk-v1")
 * ```
 *
 * @param packageNames `string[]`
 *
 * @returns `<THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig` Hardhat config decorator
 */
export const withLayerZeroArtifacts = (...packageNames: string[]) => {
    const resolvedArtifactsDirectories = packageNames
        // The first thing we do is we resolve the paths to LayerZero packages
        .map(resolvePackageDirectory)
        // Then navigate to the artifacts folder
        .map((resolvedPackagePath) => join(resolvedPackagePath, 'artifacts'))

    // We return a function that will enrich hardhat config with the external artifacts configuration
    //
    // This is a pretty standard way of enriching configuration files that leads to quite nice consumer code
    return <THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig => {
        // We'll first grab all the external artifacts already defined
        const existingArtifacts = new Set(config.external?.contracts?.flatMap(({ artifacts }) => artifacts) ?? [])

        // And only append stuff if we have something new to say
        const newArtifacts = new Set(
            resolvedArtifactsDirectories.filter((artifact) => !existingArtifacts.has(artifact))
        )
        if (newArtifacts.size === 0) {
            return config
        }

        return {
            ...config,
            external: {
                ...config.external,
                contracts: [
                    ...(config.external?.contracts ?? []),
                    {
                        artifacts: Array.from(newArtifacts),
                    },
                ],
            },
        }
    }
}
