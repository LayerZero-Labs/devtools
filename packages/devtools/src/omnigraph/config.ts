import type { Factory } from '@/types'
import type { Configurator, IOmniSDK, InferOmniEdge, InferOmniNode, OmniGraph, OmniSDKFactory } from './types'
import type { OmniTransaction } from '@/transactions/types'
import { flattenTransactions } from '@/transactions/utils'

export type CreateTransactionsFromOmniNodes<TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK> = Factory<
    [InferOmniNode<TOmniGraph>, TOmniSDK, TOmniGraph, OmniSDKFactory<TOmniSDK>],
    OmniTransaction[] | OmniTransaction | null | undefined
>

export type CreateTransactionsFromOmniEdges<TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK> = Factory<
    [InferOmniEdge<TOmniGraph>, TOmniSDK, TOmniGraph, OmniSDKFactory<TOmniSDK>],
    OmniTransaction[] | OmniTransaction | null | undefined
>

/**
 * Function that takes care of the boilerplate for node configuration functions.
 *
 * It will create an SDK for every node in the graph, then call `createTransactions`
 * with the node itself, the created SDK, the whole graph and the SDK factory.
 *
 * ```
 * const configureSomething = createConfigureNodes(async ({ config }: OmniNode<{ something: string }>, sdk) => {
 *   const something = await sdk.getSomething()
 *   if (something !== config.something) return []
 *
 *   return sdk.setSomething(config.something)
 * })
 * ```
 *
 * @template TOmniGraph
 * @template TOmniSDK
 * @param {CreateTransactionsFromOmniNodes<TOmniGraph, TOmniSDK>} createTransactions
 * @returns {Configurator<TOmniGraph, TOmniSDK>}
 */
export const createConfigureNodes =
    <TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK>(
        createTransactions: CreateTransactionsFromOmniNodes<TOmniGraph, TOmniSDK>
    ): Configurator<TOmniGraph, TOmniSDK> =>
    async (graph, createSdk) =>
        flattenTransactions(
            await Promise.all(
                graph.contracts.map(async (node) => {
                    const sdk = await createSdk(node.point)

                    return await createTransactions(node, sdk, graph, createSdk)
                })
            )
        )

/**
 * Function that takes care of the boilerplate for edge configuration functions.
 *
 * It will create an SDK for every edge (using the `from` field) in the graph, then call `createTransactions`
 * with the edge itself, the created SDK, the whole graph and the SDK factory.
 *
 * ```
 * const configureSomething = createConfigureEdges(async ({ config, vector: { to } }: OmniEdge<{ something: string }>, sdk) => {
 *   const something = await sdk.getSomethingFor(to.eid)
 *   if (something !== config.something) return []
 *
 *   return sdk.setSomething(to.eid, config.something)
 * })
 * ```
 *
 * @template TOmniGraph
 * @template TOmniSDK
 * @param {CreateTransactionsFromOmniEdges<TOmniGraph, TOmniSDK>} createTransactions
 * @returns {Configurator<TOmniGraph, TOmniSDK>}
 */
export const createConfigureEdges =
    <TOmniGraph extends OmniGraph = OmniGraph, TOmniSDK = IOmniSDK>(
        createTransactions: CreateTransactionsFromOmniEdges<TOmniGraph, TOmniSDK>
    ): Configurator<TOmniGraph, TOmniSDK> =>
    async (graph, createSdk) =>
        flattenTransactions(
            await Promise.all(
                graph.connections.map(async (edge) => {
                    const sdk = await createSdk(edge.vector.from)

                    return await createTransactions(edge, sdk, graph, createSdk)
                })
            )
        )
