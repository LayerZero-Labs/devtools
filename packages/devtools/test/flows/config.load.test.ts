import { configLoadFlow } from '@/flows/config.load'
import { importDefault, isFile, isReadable } from '@layerzerolabs/io-devtools'
import { z } from 'zod'

jest.mock('@layerzerolabs/io-devtools', () => ({
    ...jest.requireActual('@layerzerolabs/io-devtools'),
    isFile: jest.fn().mockReturnValue(false),
    isReadable: jest.fn().mockReturnValue(false),
    importDefault: jest.fn().mockRejectedValue('Not mocked'),
}))

const isFileMock = isFile as jest.Mock
const importDefaultMock = importDefault as jest.Mock
const isReadableMock = isReadable as jest.Mock

describe('flows/config.load', () => {
    beforeEach(() => {
        isFileMock.mockReset()
        importDefaultMock.mockReset()
        isReadableMock.mockReset()
    })

    describe('configLoadFlow', () => {
        it('should reject if the path is not a file', async () => {
            isFileMock.mockReturnValue(false)

            await expect(
                configLoadFlow({ configPath: './myconfig.ts', configSchema: z.unknown() })
            ).rejects.toMatchSnapshot()
        })

        it('should reject if the path is not readable', async () => {
            isFileMock.mockReturnValue(true)
            isReadableMock.mockReturnValue(false)

            await expect(
                configLoadFlow({ configPath: './myconfig.ts', configSchema: z.unknown() })
            ).rejects.toMatchSnapshot()
        })

        it('should reject if the file cannot be imported', async () => {
            isFileMock.mockReturnValue(true)
            isReadableMock.mockReturnValue(true)
            importDefaultMock.mockRejectedValue('No way')

            await expect(
                configLoadFlow({ configPath: './myconfig.ts', configSchema: z.unknown() })
            ).rejects.toMatchSnapshot()
        })

        describe('when config file exports a config', () => {
            it('should reject if the file contents do not match the schema', async () => {
                isFileMock.mockReturnValue(true)
                isReadableMock.mockReturnValue(true)
                importDefaultMock.mockResolvedValue({ bad: 'config' })

                await expect(
                    configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                ).rejects.toMatchSnapshot()
            })

            it('should resolve if the file contents match the schema', async () => {
                isFileMock.mockReturnValue(true)
                isReadableMock.mockReturnValue(true)
                importDefaultMock.mockResolvedValue({ good: 'config' })

                await expect(
                    configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                ).resolves.toMatchSnapshot()
            })
        })

        describe('when config file exports a function', () => {
            describe('when it is a synchronous function', () => {
                it('should reject if the function throws an error', async () => {
                    const configFunctionMock = jest.fn(() => {
                        throw new Error('Oh not again')
                    })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    await expect(
                        configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                    ).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should reject if the function returns an invalid config', async () => {
                    const configFunctionMock = jest.fn().mockReturnValue({ bad: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    await expect(
                        configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                    ).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should resolve if the function returns a valid config', async () => {
                    const configFunctionMock = jest.fn().mockReturnValue({ good: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    await expect(
                        configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                    ).resolves.toEqual({ good: 'config' })

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })
            })

            describe('when it is an asynchronous function', () => {
                it('should reject if the function rejects', async () => {
                    const configFunctionMock = jest.fn().mockRejectedValue(new Error('Y u do dis'))

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    await expect(
                        configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                    ).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should reject if the function resolves with an invalid config', async () => {
                    const configFunctionMock = jest.fn().mockResolvedValue({ bad: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    await expect(
                        configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                    ).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should resolve if the function returns a valid config', async () => {
                    const configFunctionMock = jest.fn().mockResolvedValue({ good: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    await expect(
                        configLoadFlow({ configPath: './myconfig.ts', configSchema: z.object({ good: z.string() }) })
                    ).resolves.toEqual({ good: 'config' })

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })
            })
        })
    })
})
