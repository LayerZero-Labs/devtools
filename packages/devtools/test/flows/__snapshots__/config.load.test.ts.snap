// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flows/config.load createConfigLoadFlow should reject if the file cannot be imported 1`] = `[Error: Unable to read config file './myconfig.ts': No way]`;

exports[`flows/config.load createConfigLoadFlow should reject if the path is not a file 1`] = `[Error: Unable to read config file './myconfig.ts'. Check that the file exists and is readable to your terminal user]`;

exports[`flows/config.load createConfigLoadFlow should reject if the path is not readable 1`] = `[Error: Unable to read config file './myconfig.ts'. Check that the file exists and is readable to your terminal user]`;

exports[`flows/config.load createConfigLoadFlow when config file exports a config should reject if the file contents do not match the schema 1`] = `
[Error: Config from file './myconfig.ts' is malformed. Please fix the following errors:

Property 'good': Required]
`;

exports[`flows/config.load createConfigLoadFlow when config file exports a config should resolve if the file contents match the schema 1`] = `
{
  "good": "config",
}
`;

exports[`flows/config.load createConfigLoadFlow when config file exports a function when it is a synchronous function should reject if the function returns an invalid config 1`] = `
[Error: Config from file './myconfig.ts' is malformed. Please fix the following errors:

Property 'good': Required]
`;

exports[`flows/config.load createConfigLoadFlow when config file exports a function when it is a synchronous function should reject if the function throws an error 1`] = `[Error: Got an exception while executing config funtion from file './myconfig.ts': Error: Oh not again]`;

exports[`flows/config.load createConfigLoadFlow when config file exports a function when it is an asynchronous function should reject if the function rejects 1`] = `[Error: Got an exception while executing config funtion from file './myconfig.ts': Error: Y u do dis]`;

exports[`flows/config.load createConfigLoadFlow when config file exports a function when it is an asynchronous function should reject if the function resolves with an invalid config 1`] = `
[Error: Config from file './myconfig.ts' is malformed. Please fix the following errors:

Property 'good': Required]
`;
